### ISSUE TITLE
Bug in Modelica.Media.Examples.MixtureGases?

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/Modelica#710]
**Modified by dietmarw on 9 Mar 2012 13:39 UTC**

Hi all,

I think Modelica.Media.Examples.MixtureGases has a bug.

There are 2 variables set by functions:

```
Real lambda1(quantity = \"ThermalConductivity\", unit = \"W/(m.K)\", min = 0.0, max = 500.0, start = 1.0, nominal = 1.0) = 
  Modelica.Media.Examples.MixtureGases.Medium1.thermalConductivity(medium1.state,1);
Real lambda2(quantity = \"ThermalConductivity\", unit = \"W/(m.K)\", min = 0.0, max = 500.0, start = 1.0, nominal = 1.0) =
  Modelica.Media.Examples.MixtureGases.Medium2.thermalConductivity(medium2.state,1);

```

Both these functions:

```
Modelica.Media.Examples.MixtureGases.Medium1.thermalConductivity
Modelica.Media.Examples.MixtureGases.Medium2.thermalConductivity
```
contain a call to function
SingleGasNasa.thermalConductivityEstimate:
```
lambdaX[i] := Modelica.Media.IdealGases.Common.SingleGasNasa.thermalConductivityEstimate(cp[i],eta[i],method);
```

The function:

```
Modelica.Media.IdealGases.Common.SingleGasNasa.thermalConductivityEstimate

```
contains:

```
lambda := if method == 1 then eta * (Cp + 1.25 * data.R) else eta * (Cp - data.R) * (1.32 + 1.77 / (-1.0 + Cp / 8.314472));

```
where "data" variable is used but in the package SingleGasNasa
"data" is \*NOT BOUND TO ANYTHING\*!

I think this can be solved by sending data[i] as input into function:

```
Modelica.Media.IdealGases.Common.SingleGasNasa.thermalConductivityEstimate

```
from function:

```
Modelica.Media.Examples.MixtureGases.Medium2.thermalConductivity
```

Any comments? Am I wrong?
Dymola can compile this because lambda1 and lambda2 are not used in the model,
but I think this is \*REALLY\* wrong as if they become used later this model will fail.

Cheers,
Adrian Pop/

----

**Reported by adrpo on 7 Mar 2012 20:51 UTC**

Hi all,

I think Modelica.Media.Examples.MixtureGases has a bug.

There are 2 variables set by functions:

```
Real lambda1(quantity = \"ThermalConductivity\", unit = \"W/(m.K)\", min = 0.0, max = 500.0, start = 1.0, nominal = 1.0) = 
Modelica.Media.Examples.MixtureGases.Medium1.thermalConductivity(medium1.state,1);
Real lambda2(quantity = \"ThermalConductivity\", unit = \"W/(m.K)\", min = 0.0, max = 500.0, start = 1.0, nominal = 1.0) = Modelica.Media.Examples.MixtureGases.Medium2.thermalConductivity(medium2.state,1);

```

Both these functions:

```
Modelica.Media.Examples.MixtureGases.Medium1.thermalConductivity
Modelica.Media.Examples.MixtureGases.Medium2.thermalConductivity
```
contain a call to function
SingleGasNasa.thermalConductivityEstimate:
```
lambdaX[i] := Modelica.Media.IdealGases.Common.SingleGasNasa.thermalConductivityEstimate(cp[i],eta[i],method);
```

The function:

```
Modelica.Media.IdealGases.Common.SingleGasNasa.thermalConductivityEstimate

```
contains:

```
lambda := if method == 1 then eta * (Cp + 1.25 * data.R) else eta * (Cp - data.R) * (1.32 + 1.77 / (-1.0 + Cp / 8.314472));

```
where "data" variable is used but in the package SingleGasNasa
"data" is \*NOT BOUND TO ANYTHING\*!

I think this can be solved by sending data[i] as input into function:

```
Modelica.Media.IdealGases.Common.SingleGasNasa.thermalConductivityEstimate

```
from function:

```
Modelica.Media.Examples.MixtureGases.Medium2.thermalConductivity
```

Any comments? Am I wrong?
Dymola can compile this because lambda1 and lambda2 are not used in the model,
but I think this is \*REALLY\* wrong as if they become used later this model will fail.

Cheers,
Adrian Pop/

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/710

**Comment by modelica-trac-importer** [comment]
**Comment by adrpo on 7 Mar 2012 20:58 UTC**

Same issue with: Modelica.Media.Examples.TestOnly.MixIdealGasAir:

```
Real lambda(quantity = \"ThermalConductivity\", unit = \"W/(m.K)\", min = 0.0, max = 500.0, start = 1.0, nominal = 1.0) =
  Modelica.Media.Examples.TestOnly.MixIdealGasAir.Medium.thermalConductivity(medium.state,1);
```
Function Modelica.Media.Examples.TestOnly.MixIdealGasAir.Medium.thermalConductivity contains a call to:
```
lambdaX[i] := Modelica.Media.IdealGases.Common.SingleGasNasa.thermalConductivityEstimate(cp[i],eta[i],method);
```
that uses "data" that has no binding!

Cheers,
Adrian Pop/

**Comment by modelica-trac-importer** [comment]
**Comment by hubertus on 7 Mar 2012 21:19 UTC**
I think you are right, after taking a first look at it. I do not agree that the variable is not used in the model: I would consider it used when it is assigned in the declaration section. When I simulate the model in Dymola, I do get the expected result, not some nonsense. Somehow Dymola uses the right data record form the data record array in Mixture gases anyways. Whether that is according to the spec or not is not clear to me. If it is wrong here, it is in wrong for mixture gases in general.

**Comment by modelica-trac-importer** [comment]
**Comment by adrpo on 7 Mar 2012 21:43 UTC**

This is even more scary. 
There's no way Dymola knows how to pass data[i] to function SingleGasNasa.thermalConductivityEstimate without some very special hack.
Eh, I'll change our Media library and pass data[i] in function thermalConductivity and see how is working.
I'll let you know how it goes. 

Cheers,
Adrian Pop/

**Comment by modelica-trac-importer** [comment]
**Comment by hubertus on 7 Mar 2012 22:34 UTC**
That should certainly work. I have not yet merged the MediaFix branch - that was just as well - and will fix it there. Let me know how it goes for you.

**Comment by modelica-trac-importer** [comment]
**Comment by adrpo on 7 Mar 2012 22:58 UTC**

Hi,

I got the two models to simulate now: 
Modelica.Media.Examples.MixtureGases.mos
Modelica.Media.Examples.TestOnly.MixIdealGasAir.mos

My changes into: Modelica/Media/IdealGases/Common/package.mo
```
Index: package.mo
===================================================================
--- package.mo	(revision 11335)
+++ package.mo	(revision 11336)
@@ -525,6 +525,7 @@
   function thermalConductivityEstimate
     "Thermal conductivity of polyatomic gases(Eucken and Modified Eucken correlation)"
     extends Modelica.Icons.Function;
+    input IdealGases.Common.DataRecord data "Ideal gas data";
     input SpecificHeatCapacity Cp "Constant pressure heat capacity";
     input DynamicViscosity eta "Dynamic viscosity";
     input Integer method(min=1,max=2)=1
@@ -554,7 +555,7 @@
   algorithm
     assert(fluidConstants[1].hasCriticalData,
     "Failed to compute thermalConductivity: For the species \"" + mediumName + "\" no critical data is available.");
-    lambda := thermalConductivityEstimate(specificHeatCapacityCp(state),
+    lambda := thermalConductivityEstimate(data, specificHeatCapacityCp(state),
       dynamicViscosity(state), method=method);
     annotation (smoothOrder=2);
   end thermalConductivity;
@@ -1357,7 +1358,7 @@
                        fluidConstants[i].acentricFactor,
                        fluidConstants[i].dipoleMoment);
     cp[i] := SingleGasNasa.cp_T(data[i],state.T);
-    lambdaX[i] :=SingleGasNasa.thermalConductivityEstimate(Cp=cp[i], eta=
+    lambdaX[i] :=SingleGasNasa.thermalConductivityEstimate(data[i],Cp=cp[i], eta=
           eta[i], method=method);
       end for;
       lambda := lowPressureThermalConductivity(massToMoleFractions(state.X,
```

Cheers,
Adrian Pop/

**Comment by modelica-trac-importer** [comment]
**Comment by jmattsson on 8 Mar 2012 08:58 UTC**
Dymola doesn't seem to find the correct data record either.

When I added the following line to thermalConductivityEstimate:
```mo
Modelica.Utilities.Streams.print("data.R = " + String(data.R));
```
and simulated MixtureGases, then I got a lot of "`data.R = 0`" in the log.

**Comment by modelica-trac-importer** [comment]
**Comment by hubertus on 8 Mar 2012 17:03 UTC**
The problem was fixed by passing data as input to SingleGasNasa.thermalConductivityEstimate in r4979. It has not been merged to the maintenance branches or trunk, so i keep the ticket open until that is done. I will wait with that until I get more feedback from other implementors on that the Media library seems ok (no further errors detected).

**Comment by modelica-trac-importer** [comment]
**Comment by kurzbach on 9 Mar 2012 13:23 UTC**
We (ITI) agree that this is clearly wrong and a bug in the library. The proposed solution seems to be reasonable, because there are many other similar functions, taking already the data record as argument.

**Comment by modelica-trac-importer** [comment]
**Modified by dietmarw on 9 Mar 2012 13:39 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 25 Apr 2012 13:41 UTC**
The proposed solution that still has a call to SingleGasNasa.thermalConductivityEstimate which is a non-intuitive structuring to me.

SingleGasNasa is partial:
Section 5.3.2 does for composite name lookup state:
"The class we look inside may not be partial in a simulation model."

This restriction should logically also apply to functions called in simulation models. 
The intent is that one does not use SingleGasNasa, but one of the packages derived from it - which clearly does not work in this case.

So in order to handle that the complete functions like cp_T, thermalConductivityEstimate, dynamicViscosityLowPressure, etc the simple solution is to implement them in another package; the functions in SingleGasNasa are then replaced by trivial functions calling those implementations.

This implies that the we can keep the interface of functions in SingleGasNasa unchanged; or make 'data' implicit in all of them - instead of having a mix of implicit and explicit passing of data in SingleGasNasa.

The proposal is something like:
```
  partial package SingleGasNasa
    constant IdealGases.Common.DataRecord data;
    ...
    function thermalConductivityEstimate
      ...
    algorithm
      lambda:=SingleGasFunctions.thermalConductivityEstimate(..., data)
    end thermalConductivityEstimate;
  end SingleGasNasa;
  partial package MixtureGasNasa
    ..
      SingleGasFunctions.thermalConductivityEstimate(...)
    ..
  end MixtureGasNasa;
  package SingleGasFunctions
     function thermalConductivityEstimate
        input IdealGases.Common.DataRecord data;
        ...
      algorithm
        lambda:=...;
      end thermalConductivityEstimate;
  end SingleGasFunctions;
```

This has the simpler rule that you do not call any function directly in SingleGasNasa (from the outside) - instead of saying that you can call some of the functions, but you have to verify  that it does not depend on the package contents.

**Comment by modelica-trac-importer** [comment]
**Modified by hubertus on 28 May 2012 22:49 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by hubertus on 10 Jun 2012 23:34 UTC**
Has been fixed with the merge of a251c81439d91417bda21fba4ef0f945a3854365

**Comment by modelica-trac-importer** [comment]
**Modified by dietmarw on 25 Jul 2012 14:49 UTC**
